PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/common/classifiedexception/ClassifiedExceptionFactory.kt

```diff
@@ -0,0 +1,44 @@
+package com.microsoft.azure.hdinsight.common.classifiedexception
+
+import com.microsoft.azure.hdinsight.common.logger.ILogger
+import org.apache.commons.lang.exception.ExceptionUtils
+import java.io.IOException
+
+const val ToolPackageSuffix: String = "com.microsoft.azure.hdinsight"
+const val EmptyLog: String = "Exception without detail message"
+
+enum class ClassifiedExceptionType {
+    USER_EXCEPTION,
+    SYS_EXCEPTION,
+    TOOL_EXCEPTION
+}
+
+object ClassifiedExceptionFactory : ILogger {
+    fun createClassifiedException(exp: Throwable?, message: String?): ClassifiedException? {
+
+        ///check whether context info is enough to create a classified exception
+        if (exp == null && message == null) {
+            log().warn("Cannot create classified exception since have not enough context")
+            return null
+        }
+
+        val rootCause: Throwable? = (exp?.cause) ?: exp
+        val msg: String = (rootCause?.message) ?: (message as String)
+
+        when {
+            isSysException(rootCause) -> return SparkSystemException(rootCause, msg)
+            isToolException(rootCause, msg) -> return SparkToolException(rootCause, msg)
+            else ->
+                return SparkUserException(rootCause, msg)
+        }
+    }
+
+    fun isSysException(rootCause: Throwable?): Boolean {
```

Comment:
We can combine `isMe()` method with every classes' factory method inheriting from `ClassifiedException`